packages/amplify-graphql-auth-transformer/src/graphql-auth-transformer.ts

```diff
@@ -716,55 +716,133 @@ export class AuthTransformer extends TransformerAuthBase implements TransformerA
-        roleDefinitions,
+        relatedModelReadRoleDefinitions,
         relatedModelObject.fields ?? [],
       );
+      // When the default redaction is false, it means no field resolver is involved
+      // Need the additional check on model level auth rules of both sides to determine the relational field redaction
+      if (!redactRelationalField) {
+        let filteredRelatedModelReadRoleDefinitions = relatedModelReadRoleDefinitions;
+        // When userpool private roles are detected, filter out the non-private userpool roles
+        if (filteredRelatedModelReadRoleDefinitions.some((r) => r.provider === 'userPools' && r.strategy === 'private')) {
+          filteredRelatedModelReadRoleDefinitions = filteredRelatedModelReadRoleDefinitions.filter(
+            (r) => !(r.provider === 'userPools' && r.strategy !== 'private'),
+          );
+        }
+        // When oidc private roles are detected, filter out the non-private oidc roles
+        if (filteredRelatedModelReadRoleDefinitions.some((r) => r.provider === 'oidc' && r.strategy === 'private')) {
+          filteredRelatedModelReadRoleDefinitions = filteredRelatedModelReadRoleDefinitions.filter(
+            (r) => !(r.provider === 'oidc' && r.strategy !== 'private'),
+          );
+        }
+        // relational field read roles are already processed with filter in parent call
+        const fieldReadRoleDefinitions = fieldRoles.map((r) => this.roleMap.get(r)!);
+        /**
+         * Compare if two auth roles are identical
+         * @param role1 auth role 1
+         * @param role2 auth role 2
+         * @returns boolean for two roles are identical
+         */
+        const isIdenticalAuthRole = (role1: RoleDefinition, role2: RoleDefinition): boolean =>
+          role1.provider === role2.provider &&
+          role1.strategy === role2.strategy &&
+          role1.static === role2.static &&
+          role1.claim === role2.claim &&
+          role1.entity === role2.entity;
+        /**
+         * Determine if the given relational field role have the same access as one of the related model auth roles
+         * Usually this means an auth role have the exact same provider, strategy, claim and entity
+         * One special case is when the provider 'userPools' or 'oidc' exists on both sides and at least one side have 'private' role
+         * @param fieldRole relational field auth role
+         * @param relatedModelRoles related model auth roles
+         * @returns boolean for the field can access both sides or not
+         */
+        const isFieldRoleHavingAccessToBothSide = (fieldRole: RoleDefinition, relatedModelRoles: RoleDefinition[]): boolean =>
+          relatedModelRoles.some(
+            (relatedRole) =>
+              isIdenticalAuthRole(relatedRole, fieldRole) ||
+              (relatedRole.provider === fieldRole.provider &&
+                (relatedRole.provider === 'userPools' || relatedRole.provider === 'oidc') &&
```

Comment:
1. Wouldn't `identityPools` be a valid provider for this case?
2. As above, put this utility in RoleDefinition & make sure we have plenty of tests for it